### PR TITLE
Reworking graphQl domains

### PR DESCRIPTION
**Pull request by JanuszJakubiec**

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/100212) / small_tests / aa754c4e17088a063caa6891b4c2d92d9c84422b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/100212/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/100212/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-13_09.07.30/lib.mongooseim.logs/run.2022-01-13_09.07.31/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/100250) / small_tests / 82d05378352fc120cf4f0108392e9b8ba64d54d4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/100250/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/100250/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-13_10.39.25/lib.mongooseim.logs/run.2022-01-13_10.39.25/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/100283) / small_tests / 1d26bf06eca63412c566b70f9b5bf8b2c07983de
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/100283/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/100283/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-13_11.33.41/lib.mongooseim.logs/run.2022-01-13_11.33.41/suite.log.html)

**Review by Premwoik — COMMENTED** [review]
Looks good to me :) I have only a few small suggestions

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/100558) / small_tests / 6947fd83569ceaf86c2cce665dbb8106c727a9cf
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/100558/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/100558/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-14_08.06.50/lib.mongooseim.logs/run.2022-01-14_08.06.50/suite.log.html)

File: src/mongoose_graphql/admin/mongoose_graphql_domain_admin_mutation.erl [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+-ignore_xref([execute/4]).
+
+-include("../mongoose_graphql_types.hrl").
+
+execute(_Ctx, admin, <<"addDomain">>, #{<<"domain">> := Domain, <<"hostType">> := HostType}) ->
+    case mongoose_domain_api:insert_domain(Domain, HostType) of
+        ok ->
+            {ok, #domain{domain = Domain, host_type = HostType}};
+        {error, _} = Err ->
+            Err
```
Comment (chrzaszcz) [inline src/mongoose_graphql/admin/mongoose_graphql_domain_admin_mutation.erl:14]: I think we will need test cases for the error conditions.

File: big_tests/tests/graphql_domain_SUITE.erl [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    ?assertEqual(#{<<"domain">> => DomainName2,
+                   <<"hostType">> => ?HOST_TYPE,
+                   <<"enabled">> => null}, ParsedResult2).
+
+disable_domain(Config) ->
+    Vars = #{domain => <<"exampleDomain">>},
+    Result = execute_auth(#{query => disable_domain_call(), variables => Vars,
+                   operationName => <<"M1">>}, Config),
+    ParsedResult = ok_result(<<"domains">>, <<"disableDomain">>, Result),
+    ?assertEqual(#{<<"domain">> => <<"exampleDomain">>, <<"enabled">> => false}, ParsedResult).
```
Comment (chrzaszcz) [inline big_tests/tests/graphql_domain_SUITE.erl:79]: Could you check that the domain is really disabled here e.g. with calling the node with RPC or using `get_domains_by_host_type`?

**Review by chrzaszcz — COMMENTED** [review]
It looks good in general, but I think it requires some tests for the error handling. If needed, I could look into it in more detail on Monday and provide some suggestions regarding errors.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/100892) / small_tests / bf71be672667199ce1bbcbdebee58898aab9caa9
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/100892/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/100892/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-17_09.12.26/lib.mongooseim.logs/run.2022-01-17_09.12.27/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/101266) / small_tests / cee004893231a30d83564310271d7d7f2724a58f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/101266/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/101266/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-18_13.13.18/lib.mongooseim.logs/run.2022-01-18_13.13.19/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/101291) / small_tests / 221f46182d2cfae9819ee338c34a887af7d40a47
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/101291/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/101291/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-18_13.18.41/lib.mongooseim.logs/run.2022-01-18_13.18.41/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/101317) / small_tests / 623f62a71d876b8869c286de80d6047ae2c7ef7f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/101317/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/101317/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-18_14.01.01/lib.mongooseim.logs/run.2022-01-18_14.01.01/suite.log.html)

File: big_tests/tests/graphql_domain_SUITE.erl [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+                   <<"enabled">> => null}, ParsedResult1),
+
+    Vars2 = #{domain => DomainName2, hostType => ?HOST_TYPE},
+    Result2 = execute_auth(#{query => create_domain_call(), variables => Vars2,
+                   operationName => <<"M1">>}, Config),
+    ParsedResult2 = ok_result(<<"domains">>, <<"addDomain">>, Result2),
+    ?assertEqual(#{<<"domain">> => DomainName2,
+                   <<"hostType">> => ?HOST_TYPE,
+                   <<"enabled">> => null}, ParsedResult2).
+
```
Comment (chrzaszcz) [inline big_tests/tests/graphql_domain_SUITE.erl:79]: I would use a helper here to make it obvious that the domain name is the only real difference here.

```suggestion
create_domain(Config) ->
    create_domain(Config, <<"exampleDomain">>),
    create_domain(Config, <<"exampleDomain2">>).

create_domain(Config, DomainName) ->
    Vars = #{domain => DomainName, hostType => ?HOST_TYPE},
    Result2 = execute_auth(#{query => create_domain_call(), variables => Vars,
                   operationName => <<"M1">>}, Config),
    ParsedResult = ok_result(<<"domains">>, <<"addDomain">>, Result),
    ?assertEqual(#{<<"domain">> => DomainName,
                   <<"hostType">> => ?HOST_TYPE,
                   <<"enabled">> => null}, ParsedResult).

```

File: big_tests/tests/graphql_domain_SUITE.erl [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+     Opts = get_listener_opts(Ep),
+     User = proplists:get_value(username, Opts),
+     Password = proplists:get_value(password, Opts),
+     execute(Ep, Body, {User, Password}).
+
+ok_result(What1, What2, {{<<"200">>, <<"OK">>}, #{<<"data">> := Data}}) ->
+   maps:get(What2, maps:get(What1, Data)).
+
+error_result(ErrorNumber, {{<<"200">>, <<"OK">>}, #{<<"errors">> := Errors}}) ->
+    lists:nth(ErrorNumber, Errors).
```
Comment (chrzaszcz) [inline big_tests/tests/graphql_domain_SUITE.erl:280]: Minor: inconsistent indentation

File: src/mongoose_graphql/admin/mongoose_graphql_domain_admin_query.erl [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    Domains = mongoose_domain_api:get_domains_by_host_type(HostType),
+    Domains2 = lists:map(fun(D) -> {ok, D} end, Domains),
+    {ok, Domains2};
+execute(_Ctx, admin, <<"domainDetails">>, #{<<"domain">> := Domain}) ->
+    case mongoose_domain_sql:select_domain(Domain) of
+        {ok, #{host_type := HostType, enabled := Enabled}} ->
+            {ok, #domain{host_type = HostType, domain = Domain,
+                         enabled = Enabled}};
+        {error, not_found} ->
+            {error, domain_not_found}
```
Comment (chrzaszcz) [inline src/mongoose_graphql/admin/mongoose_graphql_domain_admin_query.erl:19]: How about this one - I am not sure, but it looks to be unhandled by the error formatter as it is not a map?

**Review by chrzaszcz — COMMENTED** [review]
It looks good, I only found a few minor details that could be improved.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/101461) / small_tests / 005f4c510d3f44eedbbc0834d84212ea806cdb53
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/101461/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/101461/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-19_08.32.18/lib.mongooseim.logs/run.2022-01-19_08.32.19/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/101512) / small_tests / 64307da2abab8337e5956200928a754be959c16f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/101512/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/101512/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-19_08.47.47/lib.mongooseim.logs/run.2022-01-19_08.47.48/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/101589) / small_tests / c8478b8e74dea9193c024116c9798fd9c1129d90
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/101589/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/101589/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-19_09.42.53/lib.mongooseim.logs/run.2022-01-19_09.42.53/suite.log.html)

File: big_tests/tests/graphql_domain_SUITE.erl [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+                       msg
+                       domain
+                           {domain}
+                   }
+               }
+           }">>.
+
+%% Helpers
+
+execute_auth(Body, Config) ->
```
Comment (Premwoik) [inline big_tests/tests/graphql_domain_SUITE.erl:281]: You can now import `execute_auth` from `graphql_helper`, then `get_listener_opts` can be removed.

**Review by Premwoik — COMMENTED** [review]
Looks good :) Only one minor comment.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/101822) / small_tests / 2f1ea615a564e8fbbb09e6dc11858ffb1b585006
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/101822/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/101822/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-20_08.37.03/lib.mongooseim.logs/run.2022-01-20_08.37.04/suite.log.html)

**Review by Premwoik — APPROVED** [review]
great :)

File: big_tests/tests/graphql_domain_SUITE.erl [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+domain_not_found_error_formatting_after_mutation_enable_domain(Config) ->
+    DomainName = <<"UnexisitingDomain">>,
+    Vars = #{domain => DomainName},
+    Result = execute_auth(#{query => enable_domain_call(), variables => Vars,
+                   operationName => <<"M1">>}, Config),
+    domain_not_found_error_formatting(Result, DomainName, <<"enableDomain">>).
+
+domain_not_found_error_formatting_after_mutation_disable_domain(Config) ->
+    DomainName = <<"UnexisitingDomain">>,
```
Comment (arcusfelis) [inline big_tests/tests/graphql_domain_SUITE.erl:113]:
```suggestion
    DomainName = <<"UnexistingDomain">>,
```

**Review by arcusfelis — APPROVED** [review]
looks good, just one typo, please, fix before merging

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/101997) / small_tests / 2bb3006df9f6eca84e85c8f953cdd4f62fb53363
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3499/101997/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3499/101997/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-21_11.35.22/lib.mongooseim.logs/run.2022-01-21_11.35.22/suite.log.html)

**Review by chrzaszcz — APPROVED** [review]
Looks good, thanks for all the corrections along the way!

**Pull request merged**